[JIT] Add support for backend-lowered submodules

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41146 [JIT] Add support for backend-lowered submodules**

**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

Differential Revision: [D22459543](https://our.internmc.facebook.com/intern/diff/D22459543)

Commit:
[JIT] Add support for backend-lowered submodules

**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

SplitInfinity commented:
@suo This is the one that got reverted but somehow the tests are clean now without any extra changes? I don't know what's going on :(

malfet commented on test/jit/test_backends.py:
```diff
@@ -69,47 +76,129 @@ def compare_py_jit_backend(self, name, input):
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.basic_module_test = BasicModuleTest()
+        self.nested_module_test = NestedModuleTest()
```
You need to construct BasicModuleTest/NestedModuleTest with the same name
```suggestion
        self.basic_module_test = BasicModuleTest(name)
        self.nested_module_test = NestedModuleTest(name)
```

malfet commented on test/jit/test_backends.py:
```diff
@@ -69,47 +76,129 @@ def compare_py_jit_backend(self, name, input):
+        self.basic_module_test = BasicModuleTest()
+        self.nested_module_test = NestedModuleTest()
+
+    def setUp(self):
```
```suggestion
    def setUp(self):
        super().setUp(self)
```

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

Differential Revision: [D22459543](https://our.internmc.facebook.com/intern/diff/D22459543)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

Differential Revision: [D22459543](https://our.internmc.facebook.com/intern/diff/D22459543)

[ghstack-poisoned]
